executor/index_merge_reader.go

```diff
@@ -0,0 +1,738 @@
+			doneCh := make(chan error, 1)
+			doneCh <- err4Panic
+			w.resultCh <- &lookupTableTask{
+				doneCh: doneCh,
+			}
+			if err != nil {
+				err = errors.Trace(err4Panic)
+			}
+		}
+	}()
+	var chk *chunk.Chunk
+	handleOffset := -1
+	if w.tableInfo.PKIsHandle {
+		handleCol := w.tableInfo.GetPkColInfo()
+		// colums := w.tableReader.Schema().Columns
+		columns := w.tableInfo.Columns
+		for i := 0; i < len(columns); i++ {
+			if columns[i].Name.L == handleCol.Name.L {
+				handleOffset = i
+				break
+			}
+		}
+	} else {
+		return 0, errors.Errorf("cannot find the column for handle")
+	}
+
+	chk = chunk.NewChunkWithCapacity(retTypes(w.tableReader), w.maxChunkSize)
+	for {
+		handles, retChunk, err := w.extractTaskHandles(ctx, chk, handleOffset)
+		if err != nil {
+			doneCh := make(chan error, 1)
+			doneCh <- err
+			w.resultCh <- &lookupTableTask{
+				doneCh: doneCh,
+			}
+			return count, err
+		}
+		if len(handles) == 0 {
+			task := w.buildTableTask(handles, retChunk)
+			w.fetchCh <- task
+			return count, nil
+		}
+		count += int64(len(handles))
+		task := w.buildTableTask(handles, retChunk)
+		select {
+		case <-ctx.Done():
+			return count, nil
+		case <-w.finished:
+			return count, nil
+		case w.fetchCh <- task:
+		}
+	}
+}
+
+func (w *partialTableWorker) extractTaskHandles(ctx context.Context, chk *chunk.Chunk, handleOffset int) (
+	handles []int64, retChk *chunk.Chunk, err error) {
+	handles = make([]int64, 0, w.batchSize)
+	for len(handles) < w.batchSize {
+		chk.SetRequiredRows(w.batchSize-len(handles), w.maxChunkSize)
+		err = errors.Trace(w.tableReader.Next(ctx, chk))
+		if err != nil {
+			return handles, nil, err
+		}
+		if chk.NumRows() == 0 {
+			return handles, retChk, nil
+		}
+		for i := 0; i < chk.NumRows(); i++ {
+			h := chk.GetRow(i).GetInt64(handleOffset)
+			handles = append(handles, h)
+		}
+	}
+	w.batchSize *= 2
+	if w.batchSize > w.maxBatchSize {
+		w.batchSize = w.maxBatchSize
+	}
+	return handles, retChk, nil
+}
+
+func (w *partialTableWorker) buildTableTask(handles []int64, retChk *chunk.Chunk) *lookupTableTask {
+	var indexOrder map[int64]int
+	var duplicatedIndexOrder map[int64]int
+	if w.keepOrder {
+		indexOrder = make(map[int64]int, len(handles))
+		for i, h := range handles {
+			indexOrder[h] = i
+		}
+	}
+	task := &lookupTableTask{
+		handles:              handles,
+		indexOrder:           indexOrder,
+		duplicatedIndexOrder: duplicatedIndexOrder,
+		idxRows:              retChk,
+	}
+
+	task.doneCh = make(chan error, 1)
+	return task
+}
+
+func (e *IndexMergeReaderExecutor) startIndexMergeTableScanWorker(ctx context.Context, workCh <-chan *lookupTableTask) {
+	lookupConcurrencyLimit := e.ctx.GetSessionVars().IndexLookupConcurrency
+	e.tblWorkerWg.Add(lookupConcurrencyLimit)
+	for i := 0; i < lookupConcurrencyLimit; i++ {
+		worker := &indexMergeTableScanWorker{
+			workCh:         workCh,
+			finished:       e.finished,
+			buildTblReader: e.buildFinalTableReader,
+			keepOrder:      false,
```

Comment:
Is it always false now? 
If it is, how about removing this field now, which can reduce the number of lines of changes in this PR?